### PR TITLE
Improve registry overwrite mechanism

### PR DESCRIPTION
**Pull request by xrstf**
**What this PR does / why we need it**:
When mirroring images into a local registry, it's very easy to alter the RepoDigest of an image. This will lead to problems with custom registries in KKP, as an image like

```
quay.io/cilium/cilium:v1.11.10@sha256:abc123
```

gets rewritten to

```
myregistry.initech.local/cilium/cilium:v1.11.10@sha256:abc123
```

but the image on the local registry might have another repo digest.

There are supposedly ways around this ([crane](https://github.com/google/go-containerregistry/tree/main/cmd/crane)?), but until those are implemented in KKP (in the `mirror-images` command) and to not force admins to follow a special procedure when mirroring images, this PR extends our registry overwrite mechanism.

Previously we would only take the registry itself into account, but with the new code, we look at and rewrite the entire Docker image (so `quay.io/kubermatic/foo:v1.2.3` instead of just `quay.io`). This allows the rewriting mechanism to behave smarter and -- for now -- always drop the digest if a local registry is used.

In the future we might make this configurable (e.g. adding an `keepImageDigests` option somewhere), but for now we drop the digest always when changing registries. The chances of a local registry getting poisoned with bad images is much smaller than with a public registry anyway.

The signature is `func(string) (string, error)`, which made all the creators a bit more tedious to write, but my thought was that the function can be used on user-supplied input where catching the error might be important. But in all cases in KKP, the images and tags are basically hardcoded, so for now I added a `Must()` helper to panic in cases that the hardcoded values are malformed.

**What type of PR is this?**
/kind bug
/kind feature

**Does this PR introduce a user-facing change? Then add your Release Note here**:
```release-note
Repo Digests are now dropped when a docker image is changed by using the overwrite-registry mechanism. Previously kept digests could lead to issues with mirrored Docker images.
```

**Documentation**:
```documentation
NONE
```

### CODE REVIEW

File: addons/csi/digitalocean/csi-driver.yaml [hunk 1]
```diff
@@ -118,7 +118,7 @@ spec:
             - name: socket-dir
               mountPath: /var/lib/csi/sockets/pluginproxy/
         - name: csi-do-plugin
-          image: '{{ Registry "docker.io" }}/digitalocean/do-csi-plugin:v4.2.0'
+          image: '{{ Image "digitalocean/do-csi-plugin:v4.2.0" }}'
```
Comment (xmudrii) [inline addons/csi/digitalocean/csi-driver.yaml:121]: Why don't we keep `docker.io` here?

**Review by xmudrii — COMMENTED** [review]
Looks great! Just one question, it affects some other addons as well.

File: pkg/controller/user-cluster-controller-manager/resources/resources/envoy-agent/daemonset.go [hunk 2]
```diff
@@ -147,14 +159,14 @@ func getInitContainers(ip net.IP, versions kubermatic.Versions, registryWithOver
...
+	}, nil
 }
 
-func getContainers(versions kubermatic.Versions, registryWithOverwrite registry.WithOverwriteFunc) []corev1.Container {
+func getContainers(versions kubermatic.Versions, imageRewriter registry.ImageRewriter) ([]corev1.Container, error) {
 	return []corev1.Container{
 		{
 			Name:            resources.EnvoyAgentDaemonSetName,
-			Image:           fmt.Sprintf("%s/%s:%s", registryWithOverwrite(resources.RegistryDocker), envoyImageName, versions.Envoy),
+			Image:           registry.Must(imageRewriter(fmt.Sprintf("%s:%s", envoyImageName, versions.Kubermatic))),
```
Comment (xmudrii) [inline pkg/controller/user-cluster-controller-manager/resources/resources/envoy-agent/daemonset.go]: Why is this changed from `versions.Envoy` to `versions.Kubermatic`?

File: addons/csi/digitalocean/csi-driver.yaml (hunk 1 above)
Comment (xrstf) [reply addons/csi/digitalocean/csi-driver.yaml:121]: Because the function will prepend it by default anyway. To me the practice of prepending it manually was always a KKP-ism, necessary for easy search&replace registry changes. But with the new code, it's not necessary anymore, so I generally removed the docker.io/ prefix.

File: pkg/controller/user-cluster-controller-manager/resources/resources/envoy-agent/daemonset.go (hunk 2 above)
Comment (xrstf) [reply pkg/controller/user-cluster-controller-manager/resources/resources/envoy-agent/daemonset.go]: Ouh that was a copy-paste mistake.

**Review by xmudrii — APPROVED** [review]
/lgtm
/approve
